R/tax_certainty.R

```diff
@@ -0,0 +1,114 @@
+#'   "NO_`rank`_SPECIFIED", and "NA"/NA.
+#'
+#' @section Developer(s):
+#'     Bruna M. Farina
+#' @section Reviewer(s):
+#'     Lewis A. Jones, Bethany J. Allen, & William Gearty
+#'
+#' @examples
+#' # Get internal data
+#' data(tetrapods)
+#' # Summarise taxonomic certainty
+#' certainty <- tax_certainty(taxdf = tetrapods, name = "accepted_name",
+#'                            certainty = c("certain", "uncertain"),
+#'                            rank = "species", add_df = TRUE)
+#' @export
+tax_certainty <- function(taxdf = NULL, name = NULL, rank = NULL,
+                          certainty = c(1, 0),
+                          add_df = TRUE) {
+  # Error handling
+  # Check taxdf is dataframe
+  if (!is.data.frame(taxdf)) {
+      stop("`taxdf` must be a data.frame")
+  }
+  # Check for valid name input
+  if (is.null(taxdf[[name]]) ||
+      !is.character(taxdf[[name]]) ||
+      length(taxdf[[name]]) == 0) {
+    stop("`names` must be a non-empty character vector and cannot be NULL")
+  }
+  # Check for valid rank input
+  if (is.null(rank) ||
+      !rank %in% c("species", "genus", "family", "order", "class")) {
+    stop("Invalid `rank`. It must be either: 'species', 'genus', 'family',
+         'order', or 'class'")
+  }
+  # Common PBDB entry (NO_RANK_SPECIFIED)
+  not_specified <- paste0("NO_", toupper(rank), "_SPECIFIED")
+  # Assign rank for handling
+  rank <- ifelse(rank == "species", "species", "other")
+  # Create temporary taxdf to not replace original values
+  tmpdf <- taxdf
+  # Replace empty rows with NA
+  tmpdf[[name]] <- gsub(pattern = "^$|^\\s+$", replacement = "NA",
+                        x = tmpdf[[name]])
+  # Change NA to character
+  tmpdf[[name]][which(is.na(tmpdf[[name]]))] <- "NA"
+  # Define match patterns
+  match_patterns <- switch(rank,
+                          "species" = c("\\S+\\s+cf\\.\\s+\\S+",
+                                        "\\S+\\s+aff\\.\\s+\\S+",
+                                        "\\S+\\s*\\?\\s*\\S+", "sp\\.$",
+                                        "informal","\\S+\\s+ex gr\\.\\s+\\S+",
+                                        "indet\\.","incertae sedis",
+                                        "problematica", "^NA$", not_specified),
+                          "other" = c("^\\cf\\.\\s*\\S+", "^\\aff\\.\\s*\\S+",
+                                      "^\\?\\S*", "incertae sedis",
+                                      "problematica", "^NA$", not_specified))
+  # Identify taxonomic certainty
+  matches <- sapply(match_patterns, grepl, tmpdf[[name]], ignore.case = TRUE)
+  # Calculate row sums (1 = match/uncertain, 0 = no match/certain)
+  if (is.null(nrow(matches))) {
+    matches <- sum(matches)
+  } else {
+    matches <- rowSums(as.matrix.data.frame(matches))
+  }
+  # Match user definitions
+  # first element of `certainty` is certain (0 + 1), second uncertain (1 + 1)
+  matches <- matches + 1
+  classif <- certainty[matches]
```

Comment:
This method of converting the uncertainty match sums needs improving. At present, T/F values are converted to 0/1 then summed, then 1 is added to match the positions in the certainty vector. However, this means certainty is matched with a value of 1 and uncertainty with a value of 2, but when more than one uncertainty match occurs (3 or above), the returned value is NA. Maybe the most efficient way is to sum the uncertainty hits then classify based on "0" and "not 0"?